### PR TITLE
drivers: counter: Adapt to use device tree

### PR DESCRIPTION
**Pull request by nordic-krch**
Modifying counter drivers (rtc and timer) to rely completely on device tree and not on Kconfig of MDK flags.

Adapting dtsi for all SoCs and adapting test configuration.

### CODE REVIEW

File: modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c [hunk 1]
```diff
@@ -8,6 +8,10 @@
 #include <platform/nrf_802154_temperature.h>
 #include <zephyr/drivers/entropy.h>
 
+/* Ensure that counter driver for TIMER1 is not enabled. */
+BUILD_ASSERT(DT_NODE_HAS_STATUS(DT_NODELABEL(timer1), disabled),
```
Comment (jciupis) [inline modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c:12]: I don't understand why this change is needed here. Could you please explain?

File: drivers/counter/Kconfig.nrfx [hunk 2]
```diff
@@ -3,108 +3,39 @@
 
 config COUNTER_NRF_TIMER
 	bool
+	default y
```
Comment (anangl) [inline drivers/counter/Kconfig.nrfx:6]: nit
```suggestion
	def_bool y
```

File: drivers/counter/Kconfig.nrfx [hunk 3]
```diff
@@ -3,108 +3,39 @@
...
+	def_bool $(dt_nodelabel_bool_prop,rtc0,zli) || \
+		 $(dt_nodelabel_bool_prop,rtc1,zli) || \
+		 $(dt_nodelabel_bool_prop,rtc2,zli) || \
+		 $(dt_nodelabel_bool_prop,timer0,zli) || \
+		 $(dt_nodelabel_bool_prop,timer1,zli) || \
+		 $(dt_nodelabel_bool_prop,timer2,zli) || \
+		 $(dt_nodelabel_bool_prop,timer3,zli) || \
+		 $(dt_nodelabel_bool_prop,timer4,zli)
+	depends on COUNTER_NRF_RTC || COUNTER_NRF_TIMER
+	select ZERO_LATENCY_IRQS
```
Comment (anangl) [inline drivers/counter/Kconfig.nrfx:41]: I don't think this `select` is a good idea. Why not ensure instead with a build assertion in the driver that `ZERO_LATENCY_IRQS` is enabled when `zli` property is set for a given (enabled) instance?

File: drivers/counter/counter_nrfx_rtc.c [hunk 4]
```diff
@@ -670,72 +672,59 @@ static const struct counter_driver_api counter_nrfx_driver_api = {
...
 			.max_top_value = COUNTER_MAX_TOP_VALUE,		       \
-			.freq = RTC_PROP(idx, clock_frequency) /	       \
-				RTC_PROP(idx, prescaler),		       \
+			.freq = DT_INST_PROP(idx, clock_frequency) /	       \
+				DT_INST_PROP(idx, prescaler),		       \
 			.flags = COUNTER_CONFIG_INFO_COUNT_UP,		       \
-			.channels = RTC_PROP(idx, fixed_top) ?		       \
-				RTC##idx##_CC_NUM : RTC##idx##_CC_NUM - 1      \
+			.channels = DT_INST_PROP(idx, fixed_top) ?	       \
+				DT_INST_PROP(idx, cc_num) : DT_INST_PROP(idx, cc_num) - 1 \
```
Comment (anangl) [inline drivers/counter/counter_nrfx_rtc.c:712]: How about:
```suggestion
			.channels = DT_INST_PROP(idx, fixed_top)	       \
				  ? DT_INST_PROP(idx, cc_num)		       \
				  : DT_INST_PROP(idx, cc_num) - 1	       \
```

File: samples/hello_world/src/main.c [hunk 5]
```diff
@@ -8,5 +8,6 @@
 
 void main(void)
 {
+	printk("%04x\n", __builtin_bswap16(0x1234));
```
Comment (anangl) [inline samples/hello_world/src/main.c:11]: Unintentional change.

File: drivers/counter/counter_nrfx_timer.c [hunk 6]
```diff
@@ -389,83 +391,62 @@ static const struct counter_driver_api counter_nrfx_driver_api = {
...
+		return init_timer(dev, &config);					\
+	}										\
+	static struct counter_nrfx_data counter_##idx##_data;				\
+	static struct counter_nrfx_ch_data						\
+		counter##idx##_ch_data[CC_TO_ID(DT_INST_PROP(idx, cc_num))];		\
+	LOG_INSTANCE_REGISTER(LOG_MODULE_NAME, idx, CONFIG_COUNTER_LOG_LEVEL);		\
+	static const struct counter_nrfx_config nrfx_counter_##idx##_config = {		\
+		.info = {								\
+			.max_top_value = (DT_INST_PROP(idx, max_bit_width) == 32) ?	\
+					0xffffffff : 0x0000ffff,			\
```
Comment (anangl) [inline drivers/counter/counter_nrfx_timer.c:434]: Maybe `BIT64_MASK()` could be used for this?

File: drivers/counter/counter_nrfx_timer.c [hunk 7]
```diff
@@ -389,83 +391,62 @@ static const struct counter_driver_api counter_nrfx_driver_api = {
...
+			.freq = TIMER_CLOCK((NRF_TIMER_Type *)DT_INST_REG_ADDR(idx)) /	\
+					(1 << DT_INST_PROP(idx, prescaler)),		\
```
Comment (anangl) [inline drivers/counter/counter_nrfx_timer.c:436]:
```suggestion
					BIT(DT_INST_PROP(idx, prescaler)),		\
```

File: modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c (hunk 1 above)
Comment (nordic-krch) [reply modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c:12]: So there is a counter API which is generic API to be used for TIMER and RTC. The configuration for counter instances used to be split between Kconfig and DT so to use one instance you had to enable it in Kconfig.
There are cases when TIMER or RTC is used outside of counter API (system timer, radio driver or bluetooth) in that case we need to protect against using same instance twice, e.g. if radio driver is used TIMER1 cannot be used for counter.

We used to have Kconfig flags for that but after moving counter configuration to DT they are useless.

So assert checks if timer1 node is disabled when radio driver is used.

Comment (jciupis) [reply modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c:12]: > if radio driver is used TIMER1 cannot be used for counter

That's reasonable, but in my opinion it's messy to put this check in a completely unrelated platform file that implements random API for the 802.15.4 driver. If anyone creates their own implementation of this API, for example using a different PRNG, they will lose this check. Wouldn't it be better to introduce this dependency in 802.15.4 driver's Kconfig by adding something like:
```
depends on $(dt_nodelabel_has_status,timer1,disabled)
```
or similar (not sure if the syntax I used is correct)?

File: drivers/counter/Kconfig.nrfx (hunk 3 above)
Comment (nordic-krch) [reply drivers/counter/Kconfig.nrfx]: Removed. If ZLI is not enabled but used there is already a BUILD_ASSERT.

File: tests/drivers/counter/counter_basic_api/boards/nrf52840dk_nrf52811.overlay [hunk 8]
```diff
@@ -1,11 +1,14 @@
...
 
 &timer1 {
+	status = "okay";
 	prescaler = <4>;
 };
 
 &timer2 {
+	status = "okay";
 	prescaler = <4>;
 };
```
Comment (anangl) [inline tests/drivers/counter/counter_basic_api/boards/nrf52840dk_nrf52811.overlay:14]: `rtc0` still missing here and in `nrf52dk_nrf52810.overlay`.

File: tests/drivers/counter/counter_basic_api/boards/nrf52840dk_nrf52840.overlay [hunk 9]
```diff
@@ -1,23 +1,34 @@
...
 };
 
 &rtc0 {
+	status = "okay";
+	ppi-wrap;
+};
+
+&rtc2 {
+	status = "okay";
 	ppi-wrap;
```
Comment (anangl) [inline tests/drivers/counter/counter_basic_api/boards/nrf52840dk_nrf52840.overlay:33]: Previously one instance was tested with `ppi-wrap` and one without. Shouldn't we keep this scenario?

File: modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c (hunk 1 above)
Comment (jciupis) [reply modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c]: LGTM, @e-rk what do you think?

File: subsys/bluetooth/controller/CMakeLists.txt [hunk 10]
```diff
@@ -20,6 +20,18 @@ zephyr_library_sources_ifdef(
   )
 
 if(CONFIG_BT_LL_SW_SPLIT)
+  dt_nodelabel(rtc0_node NODELABEL "rtc0")
```
Comment (anangl) [inline subsys/bluetooth/controller/CMakeLists.txt:23]: This is nRF-specific so I think it should go to `ll_sw/nrf.cmake`.

File: tests/drivers/counter/counter_basic_api/boards/nrf52dk_nrf52810.overlay [hunk 11]
```diff
@@ -1,11 +1,14 @@
...
 
 &timer1 {
+	status = "okay";
 	prescaler = <4>;
 };
 
 &timer2 {
+	status = "okay";
 	prescaler = <4>;
 };
```
Comment (anangl) [inline tests/drivers/counter/counter_basic_api/boards/nrf52dk_nrf52810.overlay:14]: One last missing `rtc0`. :)

File: modules/hal_nordic/nrf_802154/CMakeLists.txt [hunk 12]
```diff
@@ -5,6 +5,13 @@ zephyr_library_named(nrf-802154-platform)
 zephyr_interface_library_named(zephyr-802154-interface)
 
 if (CONFIG_NRF_802154_RADIO_DRIVER)
+  dt_nodelabel(timer1_node NODELABEL "timer1")
+  dt_node_has_status(status_result PATH ${timer1_node} STATUS okay)
+  if (${status_result})
+	  message(FATAL_ERROR "Resource conflict. IEEE802.15.4 Radio driver requires TIMER1 but "
```
Comment (e-rk) [inline modules/hal_nordic/nrf_802154/CMakeLists.txt:11]: I might be wrong, but isn't there a policy that Kconfig/DTS errors must not result in CMake fatal errors, because this effectively locks-out the user from `menuconfig`?
On top of that, catching configuration errors in CMake doesn't seem right. Can it not be somehow achieved using the Kconfig alone?

**Review by e-rk — COMMENTED** [review]
I have one comment. The Kconfigs `NRF_HW_RTCn_RESERVED` and `NRF_HW_TIMERn_RESERVED` were removed in favor of build errors in CMake/source files. One question from me:
Would it be possible to use different `compatible` for the `counter` driver and some other special purpose of the peripheral? For example, the RTC1 can be used as either `counter` or the system timer. By default, the DTS description would designate the RTC1 as `nordic,nrf-rtc-systemtimer` with default status `okay`. This would automatically configure the RTC1 to operate as system timer without conflicting with the counter, which would have a `compatible = "nordic,nrf-rtc-counter"` instead. Similar thing could be done for `TIMER` when it is used as counter or has special purpose in `nrf-802154`.

File: modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c (hunk 1 above)
Comment (e-rk) [reply modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c]: My opinion is that having the check as @jciupis suggested would be preferable. That way we stay consistent with the rest of the system. The lack of information is the problem of Kconfig itself and I believe that it should be addressed there. If every module ended up with it's own dependency checking that is parallel to Kconfig we would end up in a mess.

Comment (nordic-krch) [reply modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c]: Ok, i pushed take 2 where i used Kconfig approach. If there is a resource conflict then compilation fails without any verbal explanation.

**Review by anangl — APPROVED** [review]

**Review by e-rk — APPROVED** [review]

**Review by cvinayak — APPROVED** [review]
